tools/src/main/java/org/apache/kafka/tools/SchemaChecker/MetadataSchemaChecker.java

```diff
@@ -0,0 +1,347 @@
+            }
+        }
+    }
+
+    private static boolean checkTaggedFieldIfTag(JsonNode origNode, JsonNode newNode) {
+        if (origNode.has("tag")) {
+            if (!newNode.has("tag")) {
+                parseNewField(newNode);
+                return true;
+            }
+            if (!newNode.has("taggedVersions")) {
+                throw new IllegalStateException("new schema tagged field is missing tagged version");
+            }
+
+
+            if (latestTag + 1 != Integer.parseInt(newNode.get("tag").asText())) {
+                throw new IllegalStateException(" tag from new schema not in numeric order, " + newNode.get("tag").asText()
+                                           + " should be "  + (latestTag + 1));
+            }
+            latestTag = Integer.parseInt(newNode.get("tag").asText());
+            String[] x = cleanUpVersionStrings(origNode.get("taggedVersions").asText());
+            latestTagVersion = Integer.parseInt(cleanUpVersionStrings(origNode.get("taggedVersions").asText())[0]);
+        } else {
+            if (newNode.has("tag")) {
+                parseNewField(newNode);
+            }
+        }
+        return false;
+    }
+
+    private static void parseNewField(JsonNode node) {
+        String[] versions = cleanUpVersionStrings(String.valueOf(node.get("versions")));
+        if (oldLatestVersion + 1 != newLatestVersion && !node.has("tag")) {
+            throw new IllegalStateException("New schemas with new fields need to be on the next version iteration");
+        }
+        if (newLatestVersion != Integer.parseInt(versions[0]) && !node.has("tag")) {
+            throw new IllegalStateException("new field version not correct");
+        }
+        if (!node.has("type") && !node.has("fields")) {
+            throw new IllegalStateException("new field requires a type if it doesn't have fields node ");
+        }
+
+        if (Integer.parseInt(versions[0]) != (oldLatestVersion + 1) && !node.has("tag")) {
+            throw new IllegalStateException("New field must be on next version");
+        }
+        if (node.has("tag")) {
+            if (!node.has("taggedVersions")) {
+                throw new IllegalStateException("new tagged field is missing tagged versions");
+            }
+            String[] taggedVersions = cleanUpVersionStrings(String.valueOf(node.get("taggedVersions")));
+
+            if (Integer.parseInt(taggedVersions[0]) != latestTagVersion + 1) {
+                throw new IllegalStateException("taggedVersion incorrect for new tagged field");
+            }
+            if (Integer.parseInt(versions[0]) < oldLatestVersion ||
+                Integer.parseInt(versions[0]) > newLatestVersion) {
+                throw new IllegalStateException("Invalid version for new tagged field");
+            }
+        }
+    }
+
+    private static String[] cleanUpVersionStrings(String string) {
+        String[] cleanedInts;
+        if (string.charAt(0) == '"') {
```

Comment:
I'm confused how a quote character could end up here, does this actually happen?

Reply:
Yes but only for some. If i remember correctly versions and nullableVersions return a string contained by " characters whereas taggedVersions is not contained. Not sure why so I just worked around it.